core/state/reader.go

```diff
@@ -298,6 +332,40 @@ func (r *multiReader) Storage(addr common.Address, slot common.Hash) (common.Has
 	return common.Hash{}, errors.Join(errs...)
 }
 
+// ContractCode implements Reader, retrieving the code associated with a particular account.
+func (r *multiReader) ContractCode(addr common.Address, codeHash common.Hash) ([]byte, error) {
+	var errs []error
+	for _, reader := range r.readers {
+		// Skip state reader as it doesn't provide contract code and
+		// always returns an error.
+		if _, ok := reader.(*trieReader); ok {
```

Comment:
This is a hack: snapshot reader can never return code. So we skip it here and only focus on trie reader.

- If the state reader returns empty result and no error the multi reader can confuse it for empty code.
- If the state reader returns error it can cause multi reader to always return an error and confuse consumer of the Reader.